### DISCUSSION TITLE
Bare metal frigate installation on docker help request

### DISCUSSION THREAD

**Discussion by rugene76** [blakeblackshear/frigate#4013] [Ask A Question]
I need some help to install frigate on a bare metal docker environment.
I am stumped trying to get frigate to run. I've done all the mappings, etc, but I get a code 500 error.
It would be great if some kind person could detail how to create a container in docker for frigate. The docs are not enough.

I was successfully running frigate on HA and decided I'd like to attach to my nas which necessitated switching the install to a bare metal docker environment. I also wanted to assure good performance over my 8 cameras. All was working fine in HA except for the lack of storage and excessive processor utilization.

I'm running on an intel platform - the same one I used with HA -  with a coral m2 and an nvidia graphics card.
Docker/portainer are installed and working on ubuntu 22.04.
I can successfully deploy containers (e.g. wyze-bridge)

I am stumped trying to get frigate to run. Using the latest stable release, I've done all the mappings, etc but I get a code 500 error.
I've read and reread the install instructions.

I'm needing more explicit help and instructions that I presume would help many others.
thanks in advance!

**Comment by NickM-27** [comment]
Please provide your docker run or docker config file and a full copy of what error you're getting.

**Reply by rugene76** [reply]
I modify the example docker run command to suit my setup and I only get code 500 from portainer.  There are no logs reported by frigate. 
I'm looking for a setup walk thru help - not debugging. 
e.g., what env to set, device mapping, runtime stuff, etc. Nothing seems to work for me.
Just seeing someone elses docker run command with similar hardware would be helpful
Clearly I'm missing something on the container configuration.

**Reply by NickM-27** [reply]
The docker run command in the docs: https://docs.frigate.video/installation#docker is all it takes and is very close to what mine looks like, so I think you may just need debugging help. Unfortunately, looking for online guides you may very well just find outdated guides (Frigate has had a lot of breaking changes over the last versions) and get more confused.

**Reply by rugene76** [reply]
apparently not in my case - I'd share my run command but I'm not sure how to back it out of portainer. but I started with this:
docker run -d \
  --name frigate \
  --restart=unless-stopped \
  --mount type=tmpfs,target=/tmp/cache,tmpfs-size=1000000000 \
  --device /dev/bus/usb:/dev/bus/usb \
  --device /dev/dri/renderD128 \
  --shm-size=64m \
  -v /home/user/mnt/hubfiles/frigate:/media/frigate \
  -v /home/user/frigate/config.yml:/config/config.yml:ro \
  -v /etc/localtime:/etc/localtime:ro \
  -e FRIGATE_RTSP_PASSWORD='password' \
  -p 5000:5000 \
  -p 1935:1935 \
  blakeblackshear/frigate:stable-amd64

**Reply by NickM-27** [reply]
There are some things with portainer that adds extra complexity. For starters seeing the full error would be helpful, 500 usually means internal server error but seeing the actual text around it would give more context

**Reply by rugene76** [reply]
do coral drivers need to installed separately in ubuntu? 
or are they in the frigate build? 
My coral device is not being detected as it was in HA.

**Reply by NickM-27** [reply]
Which coral do you have? Odds are you aren't passing it in to portainer

**Reply by rugene76** [reply]
Mini-pci coral plugged into a pci slot with an adapter.
On the same hardware running ha I used the pci:0 tag in the yml and it worked fine.
What device should I map in the docker setup? /dev/pci_0 doesn't work, nor does /dev/apex_0
Thanks

**Reply by NickM-27** [reply]
Okay yes you need to install the drivers https://coral.ai/docs/m2/get-started/#2-install-the-pcie-driver-and-edge-tpu-runtime

**Reply by NickM-27** [reply]
And it's /dev/apex_0

**Reply by rugene76** [reply]
Thanks, I'll get back at it tomorrow 🌜

**Reply by rugene76** [reply]
Does it help to use more than one coral?

**Reply by NickM-27** [reply]
What do you mean by help?

**Reply by rugene76** [reply]
Improve performance, lower CPU utilization. I happen to have two

**Reply by NickM-27** [reply]
Depends on how many cameras you have and what detect fps they're set to. In small setups it won't necessarily help but it can and it won't hurt if you're not using it for anything else

**Reply by rugene76** [reply]
so I installed the coral drivers - I moved forward at least
I mapped /dev/apex_0 to /dev/apex_0 and am using the following in the config file:
detectors:
  coral1:
    type:  edgetpu
    device: pci:0

**Reply by rugene76** [reply]
Now I see no video in the web gui and get the following error log for frigate that I think is saying ffmpeg is not present. Does ffmpeg also need to be explicitly installed? Where from?
thanks
error log:
Exception in thread Thread-4:
Traceback (most recent call last):
  File "/usr/lib/python3.9/threading.py", line 954, in _bootstrap_inner
    self.run()
  File "/opt/frigate/frigate/video.py", line 215, in run
    self.start_ffmpeg_detect()
  File "/opt/frigate/frigate/video.py", line 276, in start_ffmpeg_detect
    self.ffmpeg_detect_process = start_or_restart_ffmpeg(
  File "/opt/frigate/frigate/video.py", line 133, in start_or_restart_ffmpeg
    process = sp.Popen(
  File "/usr/lib/python3.9/subprocess.py", line 951, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.9/subprocess.py", line 1823, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'ffmpeg'
Exception in thread Thread-4:
Traceback (most recent call last):
  File "/usr/lib/python3.9/threading.py", line 954, in _bootstrap_inner
    self.run()
  File "/opt/frigate/frigate/video.py", line 215, in run
    self.start_ffmpeg_detect()
  File "/opt/frigate/frigate/video.py", line 276, in start_ffmpeg_detect
    self.ffmpeg_detect_process = start_or_restart_ffmpeg(
  File "/opt/frigate/frigate/video.py", line 133, in start_or_restart_ffmpeg
    process = sp.Popen(
  File "/usr/lib/python3.9/subprocess.py", line 951, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.9/subprocess.py", line 1823, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'ffmpeg'

**Reply by NickM-27** [reply]
This is covered in the 0.11 release notes: 
> Some deployment approaches are overwriting the PATH environment value to an old value. If you see errors like No such file or directory: 'ffmpeg', then you need to stop overriding your PATH. This happens with both Proxmox LXC and Portainer (https://github.com/blakeblackshear/frigate/issues/3223)

**Reply by rugene76** [reply]
Installed ffmpeg - no change. 
Is it really this hard?

**Reply by NickM-27** [reply]
I don't think you read that right, ffmpeg is already installed in frigate but your portainer setup is overwriting frigate's path variable. So you need to go into the portainer config and make it stop overwriting the path

**Reply by rugene76** [reply]
How do you stop overriding your PATH? Portainer is automagically putting this in my env
<img width="860" alt="image" src="https://user-images.githubusercontent.com/60796202/194099306-1a2b409e-4a6a-411a-88d6-c58eefa46875.png">

**Reply by NickM-27** [reply]
hit the delete button or potentially "remove value" then start frigate

**Reply by rugene76** [reply]
deleted path and I think I'm working.

The install is ridiculously difficult - and I consider myself somewhat technical.....
I'd definitely suggest a guide be put together fully laying out a bare metal install.
Now how do I validate that my nvidia hardware is operational?.....I'm seeing 40% cpu utilization with 2 cameras using the h264_cuvid decoder

**Reply by NickM-27** [reply]
> The install is ridiculously difficult - and I consider myself somewhat technical.....

The docs covers how to install in docker and it is quite a bit simpler that way, portainer adds a lot of difficulty and complications to the install with the benefit of having a UI.

> Now how do I validate that my nvidia hardware is operational?.....I'm seeing 40% cpu utilization with 2 cameras using the h264_cuvid decoder

Run `nvidia-smi` on your host

**Reply by rugene76** [reply]
So here is my log now - no more errors.
[2022-10-05 10:35:01] frigate.app                    INFO    : Starting Frigate (0.11.0-3846a13)
Starting migrations
[2022-10-05 10:35:01] peewee_migrate                 INFO    : Starting migrations
There is nothing to migrate
[2022-10-05 10:35:01] peewee_migrate                 INFO    : There is nothing to migrate
[2022-10-05 10:35:01] frigate.app                    INFO    : Output process started: 217
[2022-10-05 10:35:01] frigate.app                    INFO    : Camera processor started for 1_driveway: 221
[2022-10-05 10:35:01] detector.coral1                INFO    : Starting detection process: 216
[2022-10-05 10:35:01] frigate.edgetpu                INFO    : Attempting to load TPU as pci:0
[2022-10-05 10:35:01] frigate.app                    INFO    : Camera processor started for 3_frontdoor: 223
[2022-10-05 10:35:01] frigate.app                    INFO    : Capture process started for 1_driveway: 225
[2022-10-05 10:35:01] frigate.app                    INFO    : Capture process started for 3_frontdoor: 227
[2022-10-05 10:35:01] ws4py                          INFO    : Using epoll
[2022-10-05 10:35:01] frigate.edgetpu                INFO    : TPU found
[2022-10-05 10:35:01] ws4py                          INFO    : Using epoll
[2022-10-05 10:35:03] ws4py                          INFO    : Managing websocket [Local => 127.0.0.1:8082 | Remote => 127.0.0.1:51464]

**Reply by rugene76** [reply]
in docker I get nvidia-smi not found using: sudo docker exec -it frigate nvidia-smi

**Reply by NickM-27** [reply]
Did you install the nvidia drivers on your host?

**Reply by rugene76** [reply]
yes installed on host and nvidia-smi responds.
tried adding nvidia gpu support (apt-get install -y nvidia-docker2, etc..) to docker but still can't turn on the GPU switch. It gives me a null error at startup.
suggestions?

**Reply by NickM-27** [reply]
> yes installed on host and nvidia-smi responds.

So what does it show?

> tried adding nvidia gpu support (apt-get install -y nvidia-docker2, etc..) to docker but still can't turn on the GPU switch. It gives me a null error at startup.
suggestions?

You do not need to install that in docker. Have you set the `--runtime=nvidia` argument in the docker setup for frigate? This is required as well.

**Reply by rugene76** [reply]
yes, I did set the runtime to nvidia but I see little change in cpu utilization. 
When I turn on the gpu I get a null error on startup. 
If the gpu is not enabled, I do, in the container console, get this result to ffmpeg -decoders | grep cuvid
V..... av1_cuvid            Nvidia CUVID AV1 decoder (codec av1)
 V..... h264_cuvid           Nvidia CUVID H264 decoder (codec h264)
 V..... hevc_cuvid           Nvidia CUVID HEVC decoder (codec hevc)
 V..... mjpeg_cuvid          Nvidia CUVID MJPEG decoder (codec mjpeg)
 V..... mpeg1_cuvid          Nvidia CUVID MPEG1VIDEO decoder (codec mpeg1video)
 V..... mpeg2_cuvid          Nvidia CUVID MPEG2VIDEO decoder (codec mpeg2video)
 V..... mpeg4_cuvid          Nvidia CUVID MPEG4 decoder (codec mpeg4)
 V..... vc1_cuvid            Nvidia CUVID VC1 decoder (codec vc1)
 V..... vp8_cuvid            Nvidia CUVID VP8 decoder (codec vp8)
 V..... vp9_cuvid            Nvidia CUVID VP9 decoder (codec vp9)

does this mean its working?
 I don't get any response to the "docker exec -it frigate nvidia-smi" command in the os. I get exec not found in path.

**Reply by NickM-27** [reply]
Please just take a screenshot of the results when frigate is running and you run `nvidia-smi` on the host (not using docker exec). That will tell if it's working or not.

**Reply by rugene76** [reply]
I've turned on 7 camera and it kinda working, very spotty video. The log is filled with this kinda of stuff:
  File "/usr/lib/python3.9/runpy.py", line 87 in _run_code
  File "/usr/lib/python3.9/runpy.py", line 197 in _run_module_as_main
Fatal Python error: Bus error

Thread 0x00007fa0596ca700 (most recent call first):
  File "/usr/lib/python3.9/threading.py", line 312 in wait
  File "/usr/lib/python3.9/multiprocessing/queues.py", line 233 in _feed
  File "/usr/lib/python3.9/threading.py", line 892 in run
  File "/usr/lib/python3.9/threading.py", line 954 in _bootstrap_inner
  File "/usr/lib/python3.9/threading.py", line 912 in _bootstrap

**Reply by NickM-27** [reply]
The docs cover this error https://docs.frigate.video/faqs#fatal-python-error-bus-error

**Reply by rugene76** [reply]
the documentation is not clear - where is the shm size changed? /etc/docker/daemon.json does not look right. 
Where in portainer? sysctls doesn't seem to work, gives a code 500 on start....nor does adding an env variable

**Reply by rugene76** [reply]
I get this in the container console:
root@1a52443388de:/opt/frigate# df -h | grep shm
shm                         64M   39M   26M  61% /dev/shm
so how do you change it?

**Reply by NickM-27** [reply]
In portainer I think you add it as an argument `--shm-size=1G` (for example)

The docs cover general docker, we can't make assumptions about what type of docker installation you're running and listing each of the many variations would be too long / potentially create more confusion.

**Comment by rugene76** [comment]
well I give up - the docker install seems to be a mess. Among other things, the live streaming randomly does not work, object detection is spotty, and the install process is way too complicated and poorly documented. Maybe if you're a developer on the project that's ok, but for even a technically savvy user, frigate is far from ready for a general population.
I guess I'll go back to my HA based install and live with the lack of nas support or consider other nvr options

**Reply by NickM-27** [reply]
No where in the docs does it recommend using portainer, and it actually says native docker is the recommended approach. That is also what the docs are based on.

Also just to be clear, HA addon is the docker install just with a few things preconfigured, running in HA addon is the exact same code as docker install.

**Reply by rugene76** [reply]
I'd suggest figuring out portainer - many users depend on it.
I understand the HA situation but the site recommends bare metal docker. Portainer should be included. To ignore it further reduces your audience.

**Reply by mimik13** [reply]
I have to agree with rugene76. Running via Portainer should be supported.

**Reply by NickM-27** [reply]
> I have to agree with rugene76. Running via Portainer should be supported.

Running via portainer is fully supported just not recommended, we have many users that do it.
